src/browser/webgl.rs

```diff
@@ -0,0 +1,332 @@
+        .as_ref(),
+      1,
+    );
+
+    self.context.draw_arrays(
+      WebGl2RenderingContext::TRIANGLES,
+      0,
+      self.length.try_into()?,
+    );
+
+    self.source.set(self.source.get() ^ 1);
+
+    Ok(())
+  }
+
+  fn create_program(
+    gl: &WebGl2RenderingContext,
+    descriptions: Vec<ShaderDescription>,
+  ) -> Result<WebGlProgram> {
+    let program = gl.create_program().ok_or("Failed to create program")?;
+
+    descriptions
+      .iter()
+      .map(|desc| Self::create_shader(gl, desc.clone()).unwrap())
+      .for_each(|shader| gl.attach_shader(&program, &shader));
+
+    gl.link_program(&program);
+
+    if !gl.get_program_parameter(&program, WebGl2RenderingContext::LINK_STATUS) {
+      return Err(
+        gl.get_program_info_log(&program)
+          .ok_or("Failed to get program log info")?
+          .into(),
+      );
+    }
+
+    gl.use_program(Some(&program));
+
+    Ok(program)
+  }
+
+  fn create_shader(
+    gl: &WebGl2RenderingContext,
+    description: ShaderDescription,
+  ) -> Result<WebGlShader> {
+    let shader = gl
+      .create_shader(description.shader_type)
+      .ok_or("Failed to create shader")?;
+
+    gl.shader_source(&shader, description.code.trim());
+    gl.compile_shader(&shader);
+
+    if !gl.get_shader_parameter(&shader, WebGl2RenderingContext::COMPILE_STATUS) {
+      return Err(
+        gl.get_shader_info_log(&shader)
+          .ok_or("Failed to get shader info log")?
+          .into(),
+      );
+    }
+
+    Ok(shader)
+  }
+
+  fn create_texture(gl: &WebGl2RenderingContext) -> Result<WebGlTexture> {
+    let canvas = gl
+      .canvas()
+      .ok_or("Failed to get canvas off of WebGL context")?
+      .cast::<HtmlCanvasElement>()?;
+
+    let texture = gl.create_texture().ok_or("Failed to create texture")?;
+
+    gl.bind_texture(WebGl2RenderingContext::TEXTURE_2D, Some(&texture));
+
+    gl.tex_image_2d_with_i32_and_i32_and_i32_and_format_and_type_and_html_canvas_element(
+      WebGl2RenderingContext::TEXTURE_2D,
+      0,
+      WebGl2RenderingContext::RGBA.try_into()?,
+      canvas.width().try_into()?,
+      canvas.height().try_into()?,
+      0,
+      WebGl2RenderingContext::RGBA,
+      WebGl2RenderingContext::UNSIGNED_BYTE,
+      &canvas,
+    )
+    .map_err(JsValueError)?;
+
+    gl.tex_parameteri(
+      WebGl2RenderingContext::TEXTURE_2D,
+      WebGl2RenderingContext::TEXTURE_MIN_FILTER,
+      WebGl2RenderingContext::LINEAR.try_into()?,
+    );
+
+    gl.tex_parameteri(
+      WebGl2RenderingContext::TEXTURE_2D,
+      WebGl2RenderingContext::TEXTURE_WRAP_S,
+      WebGl2RenderingContext::CLAMP_TO_EDGE.try_into()?,
+    );
+
+    gl.tex_parameteri(
+      WebGl2RenderingContext::TEXTURE_2D,
+      WebGl2RenderingContext::TEXTURE_WRAP_T,
+      WebGl2RenderingContext::CLAMP_TO_EDGE.try_into()?,
+    );
+
+    Ok(texture)
+  }
+
+  fn setup_triangles(gl: &WebGl2RenderingContext, program: &WebGlProgram) -> Result<usize> {
+    let vertex_data = vec![
```

Comment:
Since the third component is always `0`, we can omit them and upload a `vec2`.

Reply:
I also think for clarity this can be an associated constant, like:

```
const VERTICES: &[f64] = &[
  -1.0, -1.0,
  1.0, -1.0,
  1.0, 1.0,
  1.0, 1.0,
  -1.0, 1.0,
  -1.0, -1.0,
];
```

Then we can also avoid having to save the length, and use `Self::VERTICES.len() / 2` instead.